### PR TITLE
added E2E test for TaskList

### PR DESCRIPTION
**Pull request by NileshPatel17**
added e2e test for #960

### CODE REVIEW

**Comment by NileshPatel17** [comment]
@herflis Please review and let me know if any changes are required.

File: apps/sensenet/cypress/integration/task-list/task-list.spec.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    cy.get('[data-test="drawer-menu-item-Content"]').click()
+    cy.get('[data-test="menu-item-IT Workspace"]').click()
+    cy.get('[data-test="add-button"').click()
+    cy.get('[data-test="listitem-Task list"')
+      .click()
+      .then(() => {
+        cy.get('#DisplayName').type(taskName)
+        cy.get('#Name').type(taskName)
+        cy.contains('Submit').click()
+        cy.get(`[data-test="table-cell-${taskName}"]`).should('have.text', taskName)
```
Comment (herflis) [inline apps/sensenet/cypress/integration/task-list/task-list.spec.ts:20]: Steps 8 and 9 are missing

File: apps/sensenet/cypress/integration/task-list/task-list.spec.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  it('Task should be edited', () => {
+    const newTaskName = `${taskName}-edited`
+    cy.get('[data-test="drawer-menu-item-Content"]').click()
+    cy.get('[data-test="menu-item-IT Workspace"]').click()
+    cy.get(`[data-test="table-cell-${taskName}"]`).rightclick()
+    cy.get('[data-test="contextmenu-edit"]')
+      .click()
+      .then(() => {
+        cy.get('#DisplayName').type('-edited')
+        cy.get('#Name').type('-edited')
```
Comment (herflis) [inline apps/sensenet/cypress/integration/task-list/task-list.spec.ts:33]:
```suggestion
```

File: apps/sensenet/cypress/integration/task-list/task-list.spec.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+  it('Task should be edited', () => {
+    const newTaskName = `${taskName}-edited`
+    cy.get('[data-test="drawer-menu-item-Content"]').click()
+    cy.get('[data-test="menu-item-IT Workspace"]').click()
+    cy.get(`[data-test="table-cell-${taskName}"]`).rightclick()
+    cy.get('[data-test="contextmenu-edit"]')
+      .click()
+      .then(() => {
+        cy.get('#DisplayName').type('-edited')
```
Comment (herflis) [inline apps/sensenet/cypress/integration/task-list/task-list.spec.ts:32]: The test case is about to change the DisplayName to "Changed Test Task List"

File: apps/sensenet/cypress/integration/task-list/task-list.spec.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        cy.get('#Name').type('-edited')
+        cy.contains('Submit').click()
+        cy.get(`[data-test="table-cell-${newTaskName}"]`).should('have.text', newTaskName)
```
Comment (herflis) [inline apps/sensenet/cypress/integration/task-list/task-list.spec.ts:35]: There is no check for "Test Task List" -> it should not be in the list

File: apps/sensenet/cypress/integration/task-list/task-list.spec.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      })
+  })
+  it('Task should be deleted', () => {
+    const taskToBeDeleted = `${taskName}-edited`
+    cy.get('[data-test="drawer-menu-item-Content"]').click()
+    cy.get('[data-test="menu-item-IT Workspace"]').click()
+    cy.get(`[data-test="table-cell-${taskToBeDeleted}"]`).rightclick()
+    cy.get('[data-test="contextmenu-delete"]')
```
Comment (herflis) [inline apps/sensenet/cypress/integration/task-list/task-list.spec.ts:43]: Step3 - Tick permanently is missing

**Review by herflis — CHANGES_REQUESTED** [review]

File: apps/sensenet/cypress/integration/task-list/task-list.spec.ts (hunk 1 above)
Comment (NileshPatel17) [reply apps/sensenet/cypress/integration/task-list/task-list.spec.ts:20]: step 8: done
step 9: There is no item "System Folder".
Screenshot 2020-10-27 at 10 28 40 AM

File: apps/sensenet/cypress/integration/task-list/task-list.spec.ts (hunk 2 above)
Comment (NileshPatel17) [reply apps/sensenet/cypress/integration/task-list/task-list.spec.ts]: removed

File: apps/sensenet/cypress/integration/task-list/task-list.spec.ts (hunk 4 above)
Comment (NileshPatel17) [reply apps/sensenet/cypress/integration/task-list/task-list.spec.ts:46]: Done

File: apps/sensenet/cypress/integration/task-list/task-list.spec.ts (hunk 5 above)
Comment (NileshPatel17) [reply apps/sensenet/cypress/integration/task-list/task-list.spec.ts:55]: Done

**Comment by NileshPatel17** [comment]
@herflis Please review the changes..

File: apps/sensenet/cypress/integration/task-list/task-list.spec.ts (hunk 1 above)
Comment (pusztaienike) [reply apps/sensenet/cypress/integration/task-list/task-list.spec.ts:20]: Yeah, it is true. Task should be the only element in this case. 👍

File: apps/sensenet/cypress/integration/task-list/task-list.spec.ts (hunk 3 above)
Comment (NileshPatel17) [reply apps/sensenet/cypress/integration/task-list/task-list.spec.ts:44]: This is also done.

**Comment by pusztaienike** [comment]
Dear @NileshPatel17,

I have checked your PR and the test has failed on my machine.
Here are my reccomodatoins:

1. After creating the content "Test Task List" a notifier (snackbar) appears with the following text:
"The content <Name> has been created succesfully". Sometimes it can hide something from the screen so I suggest you to close the notifier before the next step: `cy.get('[data-test="snackbar-close"]').click()`
For this you need changes on notificationComponent
image

2. In Step8: **"Click on Test Task List in the tree"** You used: `cy.get(`[data-test=menu-item-${taskName}]`).click()` command, but there is no data-test=menu-item attribute in our code. It is also a good solution to double-click on the content in the grid (or click once in the tree). So I suggest to change it to: `cy.get(`[data-test="table-cell-${taskName}"]`).dblclick()`

3. Your solution to check the AddNewButton dropdown elements is not working for me: 
```
 const expectedMenuItems = ['Task']
            cy.get('[role="presentation"] li')
              .should('have.length', expectedMenuItems.length)
              .each(($el) => {
                expect(expectedMenuItems).to.include($el.text())
              })
```
I suggest you to change it to
```
 const expetcedMenuItems = ['Task']
            cy.get('[data-test="list-items"]')
              .should('have.length', expetcedMenuItems.length)
              .each(($span) => {
                const text = $span.text()
                if (text) {
                  expect(expetcedMenuItems).to.include(text)
                }
              })
```
For this you will need to change AddButton.tsx

image

4. Another thing:
After checking the dropdown you have to click away to close it. But Material-ui has a layer of the top of everything when renders a Dropdown so it might cause errors and warning if you try to click hidden element. You can use click({force: true}) or simply click one element from the dropdown cause in this case it does not matter. The next step will be click on Content menu.

e.g.:`cy.get('[data-test="list-items"]').first().click()`


Here are my changes: I tried with this code and it was working fine.

image

File: apps/sensenet/cypress/integration/task-list/task-list.spec.ts (hunk 3 above)
Comment (pusztaienike) [reply apps/sensenet/cypress/integration/task-list/task-list.spec.ts:44]: It is still a finding

**Comment by NileshPatel17** [comment]
@pusztaienike I am having trouble running e2e test. app is being redirected to login page.
Screenshot 2020-10-30 at 10 54 09 AM

here is my cypress.json config:-
```
{
  "baseUrl": "http://localhost:8080",
  "chromeWebSecurity": false,
  "defaultCommandTimeout": 12000,
  "env": {
     "repoUrl": "https://sn2387786740.sensenet.cloud",
     "identityServer": "https://sn2387786740-is.sensenet.cloud",
    "users": {
      "admin": {
        "clientId": "businesscat",
        "clientSecret": "LHIZ58581047EBVK",
        "id": "/Root/IMS/Public('businesscat')"
      },
      "developer": {
        "clientId": "devdog",
        "clientSecret": "LHIZ58581047EBVK",
        "id": "/Root/IMS/Public('devdog')"
      }
    }
  }
}
```
Please help

**Comment by pusztaienike** [comment]
> @pusztaienike I am having trouble running e2e test. app is being redirected to login page.
> Screenshot 2020-10-30 at 10 54 09 AM
> 
> here is my cypress.json config:-
> 
> ```
> {
>   "baseUrl": "http://localhost:8080",
>   "chromeWebSecurity": false,
>   "defaultCommandTimeout": 12000,
>   "env": {
>      "repoUrl": "https://sn2387786740.sensenet.cloud",
>      "identityServer": "https://sn2387786740-is.sensenet.cloud",
>     "users": {
>       "admin": {
>         "clientId": "businesscat",
>         "clientSecret": "LHIZ58581047EBVK",
>         "id": "/Root/IMS/Public('businesscat')"
>       },
>       "developer": {
>         "clientId": "devdog",
>         "clientSecret": "LHIZ58581047EBVK",
>         "id": "/Root/IMS/Public('devdog')"
>       }
>     }
>   }
> }
> ```
> 
> Please help

@NileshPatel17 :
We had a change in our configuration and it seems your branch is out-of-date with our develop branch. Could you please update your branch with the actual developer branch? It will solve your problem. Thank you :)

**Comment by NileshPatel17** [comment]
@pusztaienike  alright. it worked post syncing develop branch.. Changes are done as requested but tests are not getting passed.
Screenshot 2020-10-30 at 5 08 01 PM

**Comment by NileshPatel17** [comment]
@pusztaienike requested changes done.

**Pull request merged**